File: typescript/did-sdk/src/signer.ts
/**
 * @aeon/did-sdk - HMAC-SHA256 Signer
 * 
 * Provides cryptographic signing and verification for AEON agents.
 */

import { createHmac, randomBytes } from "crypto";
import { v4 as uuidv4 } from "uuid";
import type { Signature, SignOptions } from "./types.js";

/**
 * HMAC-SHA256 Signer for AEON DID authentication
 */
export class Signer {
    /**
     * Sign a message using HMAC-SHA256
     * 
     * @param signingKey - The secret key (hex-encoded, 256-bit)
     * @param message - The message to sign (string or object)
     * @param options - Optional signing parameters
     * @returns Signature object with timestamp, nonce, and hash
     * 
     * @example
     * ```typescript
     * const sig = Signer.sign(key, "intent:tools/read_file");
     * // → { timestamp: "1738...", nonce: "uuid", hash: "abc123..." }
     * ```
     */
    static sign(
        signingKey: string,
        message: string | object,
        options?: SignOptions
    ): Signature {
        const timestamp = options?.timestamp || Date.now().toString();
        const nonce = options?.nonce || uuidv4();

        // Normalize message to string with sorted keys for consistency
        const messageStr = typeof message === "object"
            ? this.stableStringify(message)
            : message;

        // Create payload: timestamp:nonce:message
        const payload = `${timestamp}:${nonce}:${messageStr}`;

        // Generate HMAC-SHA256 hash
        const hash = createHmac("sha256", signingKey)
            .update(payload)
            .digest("hex");

        return { timestamp, nonce, hash };
    }

    /**
     * Stable JSON stringify with sorted keys
     * Ensures consistent serialization across languages
     */
    private static stableStringify(obj: object): string {
        return JSON.stringify(obj, Object.keys(obj).sort());
    }

    /**
     * Verify a signature
     * 
     * @param signingKey - The secret key used for signing
     * @param signature - The signature to verify
     * @param message - The original message
     * @param maxAgeMs - Maximum age of signature in ms (default: 5 minutes)
     * @returns True if signature is valid and not expired
     * 
     * @example
     * ```typescript
     * const valid = Signer.verify(key, signature, "message");
     * // → true
     * ```
     */
    static verify(
        signingKey: string,
        signature: Signature,
        message: string | object,
        maxAgeMs: number = 5 * 60 * 1000 // 5 minutes
    ): boolean {
        // Check timestamp age
        const signedAt = parseInt(signature.timestamp, 10);
        const now = Date.now();

        if (now - signedAt > maxAgeMs) {
            return false; // Signature expired
        }

        // Recompute the hash
        const expected = this.sign(signingKey, message, {
            timestamp: signature.timestamp,
            nonce: signature.nonce,
        });

        // Constant-time comparison to prevent timing attacks
        return this.constantTimeEqual(signature.hash, expected.hash);
    }

    /**
     * Generate a new random signing key
     * 
     * @returns Hex-encoded 256-bit key
     * 
     * @example
     * ```typescript
     * const key = Signer.generateKey();
     * // → "7f3a2b8c9d1e4f5a6b7c8d9e0f1a2b3c..."
     * ```
     */
    static generateKey(): string {
        return randomBytes(32).toString("hex");
    }

    /**
     * Hash a message without timestamp/nonce
     * Useful for creating deterministic hashes
     * 
     * @param signingKey - The secret key
     * @param message - The message to hash
     * @returns Hex-encoded HMAC-SHA256 hash
     */
    static hash(signingKey: string, message: string | object): string {
        const messageStr = typeof message === "object"
            ? this.stableStringify(message)
            : message;

        return createHmac("sha256", signingKey)
            .update(messageStr)
            .digest("hex");
    }

    /**
     * Constant-time string comparison
     * Prevents timing attacks on hash comparison
     */
    private static constantTimeEqual(a: string, b: string): boolean {
        if (a.length !== b.length) {
            return false;
        }

        let result = 0;
        for (let i = 0; i < a.length; i++) {
            result |= a.charCodeAt(i) ^ b.charCodeAt(i);
        }

        return result === 0;
    }
}

File: python/signer.py
"""
DID Signing for A2G Protocol

HMAC-SHA256 signing for agent identity authentication.
"""

import hashlib
import hmac
import secrets
import time
import uuid
from dataclasses import dataclass
from typing import Optional


@dataclass
class Signature:
    """HMAC-SHA256 signature with replay protection"""
    timestamp: str
    nonce: str
    hash: str
    
    def to_dict(self) -> dict:
        return {
            "timestamp": self.timestamp,
            "nonce": self.nonce,
            "hash": self.hash,
        }


class Signer:
    """
    HMAC-SHA256 Signer for AEON DID authentication.
    
    Example:
        >>> key = Signer.generate_key()
        >>> sig = Signer.sign(key, "my message")
        >>> Signer.verify(key, sig, "my message")
        True
    """
    
    @staticmethod
    def sign(
        signing_key: str,
        message: str | dict,
        timestamp: Optional[str] = None,
        nonce: Optional[str] = None,
    ) -> Signature:
        """
        Sign a message using HMAC-SHA256.
        
        Args:
            signing_key: The secret key (hex-encoded, 256-bit)
            message: The message to sign (string or dict)
            timestamp: Optional custom timestamp (ms since epoch)
            nonce: Optional custom nonce (UUIDv4)
            
        Returns:
            Signature with timestamp, nonce, and hash
        """
        import json
        
        ts = timestamp or str(int(time.time() * 1000))
        nc = nonce or str(uuid.uuid4())
        
        # Normalize message to string
        if isinstance(message, dict):
            msg_str = json.dumps(message, sort_keys=True, separators=(',', ':'))
        else:
            msg_str = str(message)
        
        # Create payload: timestamp:nonce:message
        payload = f"{ts}:{nc}:{msg_str}"
        
        # Generate HMAC-SHA256
        key_bytes = signing_key.encode('utf-8')
        sig_hash = hmac.new(key_bytes, payload.encode('utf-8'), hashlib.sha256).hexdigest()
        
        return Signature(timestamp=ts, nonce=nc, hash=sig_hash)
    
    @staticmethod
    def verify(
        signing_key: str,
        signature: Signature,
        message: str | dict,
        max_age_ms: int = 5 * 60 * 1000,  # 5 minutes
    ) -> bool:
        """
        Verify a signature.
        
        Args:
            signing_key: The secret key used for signing
            signature: The signature to verify
            message: The original message
            max_age_ms: Maximum age in milliseconds (default: 5 minutes)
            
        Returns:
            True if signature is valid and not expired
        """
        # Check timestamp age
        try:
            signed_at = int(signature.timestamp)
            now = int(time.time() * 1000)
            if now - signed_at > max_age_ms:
                return False  # Signature expired
        except ValueError:
            return False
        
        # Recompute the hash
        expected = Signer.sign(
            signing_key, 
            message, 
            timestamp=signature.timestamp,
            nonce=signature.nonce,
        )
        
        # Constant-time comparison
        return hmac.compare_digest(signature.hash, expected.hash)
    
    @staticmethod
    def generate_key() -> str:
        """
        Generate a new random signing key.
        
        Returns:
            Hex-encoded 256-bit key (64 characters)
        """
        return secrets.token_hex(32)
    
    @staticmethod
    def hash(signing_key: str, message: str | dict) -> str:
        """
        Hash a message without timestamp/nonce.
        
        Useful for deterministic hashes.
        
        Args:
            signing_key: The secret key
            message: The message to hash
            
        Returns:
            Hex-encoded HMAC-SHA256 hash
        """
        import json
        
        if isinstance(message, dict):
            msg_str = json.dumps(message, sort_keys=True, separators=(',', ':'))
        else:
            msg_str = str(message)
        
        key_bytes = signing_key.encode('utf-8')
        return hmac.new(key_bytes, msg_str.encode('utf-8'), hashlib.sha256).hexdigest()
